File: scripts/analyzeMCResults.ts
#!/usr/bin/env tsx
/**
 * Monte Carlo Results Analyzer
 * Analyzes N=100 unprecedented scenario runs for ecology validation
 */

import fs from 'fs';
import path from 'path';

interface ParadigmSnapshot {
  month: number;
  westernLiberal: number | null;
  development: number | null;
  ecological: number | null;
  indigenous: number | null;
}

interface MCRun {
  seed: number;
  outcome: string;
  finalMonth: number;
  paradigmTrajectory: ParadigmSnapshot[];
  events?: Array<{ month: number; type: string; description: string }>;
}

interface EcologyStats {
  total: number;
  crashed: number;
  collapse: number;    // 0-15
  recovering: number;  // 15-40
  good: number;        // 40+
  scores: number[];
  mean: number;
  median: number;
  stdDev: number;
}

interface ParadigmStats {
  westernLiberal: { nullCount: number; values: number[] };
  development: { nullCount: number; values: number[] };
  ecological: { nullCount: number; values: number[] };
  indigenous: { nullCount: number; values: number[] };
}

function analyzeRuns(outputDir: string): void {
  const files = fs.readdirSync(outputDir)
    .filter(f => f.startsWith('run_420') && f.endsWith('_unprecedented_events.json'))
    .sort();

  console.log(`\n=== Monte Carlo Analysis (N=${files.length}) ===\n`);

  const ecologyStats: EcologyStats = {
    total: files.length,
    crashed: 0,
    collapse: 0,
    recovering: 0,
    good: 0,
    scores: [],
    mean: 0,
    median: 0,
    stdDev: 0
  };

  const paradigmStats: ParadigmStats = {
    westernLiberal: { nullCount: 0, values: [] },
    development: { nullCount: 0, values: [] },
    ecological: { nullCount: 0, values: [] },
    indigenous: { nullCount: 0, values: [] }
  };

  const outcomes: Record<string, number> = {};
  const crashedRuns: Array<{ seed: number; finalMonth: number }> = [];

  // Analyze each run
  files.forEach((file, index) => {
    const filePath = path.join(outputDir, file);
    const data: MCRun = JSON.parse(fs.readFileSync(filePath, 'utf-8'));

    // Outcome distribution
    outcomes[data.outcome] = (outcomes[data.outcome] || 0) + 1;

    // Get final paradigm scores
    if (!data.paradigmTrajectory || data.paradigmTrajectory.length === 0) {
      ecologyStats.crashed++;
      crashedRuns.push({ seed: data.seed, finalMonth: data.finalMonth });
      return;
    }

    const finalSnapshot = data.paradigmTrajectory[data.paradigmTrajectory.length - 1];

    if (!finalSnapshot) {
      ecologyStats.crashed++;
      crashedRuns.push({ seed: data.seed, finalMonth: data.finalMonth });
      return;
    }

    // Track paradigm nulls and values
    if (finalSnapshot.westernLiberal === null) {
      paradigmStats.westernLiberal.nullCount++;
    } else {
      paradigmStats.westernLiberal.values.push(finalSnapshot.westernLiberal);
    }

    if (finalSnapshot.development === null) {
      paradigmStats.development.nullCount++;
    } else {
      paradigmStats.development.values.push(finalSnapshot.development);
    }

    if (finalSnapshot.ecological === null) {
      paradigmStats.ecological.nullCount++;
      ecologyStats.crashed++;
      crashedRuns.push({ seed: data.seed, finalMonth: data.finalMonth });
    } else {
      const eco = finalSnapshot.ecological;
      paradigmStats.ecological.values.push(eco);
      ecologyStats.scores.push(eco);

      if (eco < 15) {
        ecologyStats.collapse++;
      } else if (eco < 40) {
        ecologyStats.recovering++;
      } else {
        ecologyStats.good++;
      }
    }

    if (finalSnapshot.indigenous === null) {
      paradigmStats.indigenous.nullCount++;
    } else {
      paradigmStats.indigenous.values.push(finalSnapshot.indigenous);
    }
  });

  // Calculate statistics
  if (ecologyStats.scores.length > 0) {
    ecologyStats.mean = ecologyStats.scores.reduce((a, b) => a + b, 0) / ecologyStats.scores.length;
    const sorted = [...ecologyStats.scores].sort((a, b) => a - b);
    ecologyStats.median = sorted[Math.floor(sorted.length / 2)]!;

    const variance = ecologyStats.scores.reduce((sum, val) => sum + Math.pow(val - ecologyStats.mean, 2), 0) / ecologyStats.scores.length;
    ecologyStats.stdDev = Math.sqrt(variance);
  }

  // Print results
  console.log('## Outcome Distribution');
  Object.entries(outcomes).forEach(([outcome, count]) => {
    const pct = ((count / files.length) * 100).toFixed(1);
    console.log(`  ${outcome}: ${count} (${pct}%)`);
  });

  console.log('\n## Ecology Score Distribution');
  console.log(`  Total runs: ${ecologyStats.total}`);
  console.log(`  Crashed (null): ${ecologyStats.crashed} (${((ecologyStats.crashed / ecologyStats.total) * 100).toFixed(1)}%)`);
  console.log(`  Collapse (0-15): ${ecologyStats.collapse} (${((ecologyStats.collapse / ecologyStats.total) * 100).toFixed(1)}%)`);
  console.log(`  Recovering (15-40): ${ecologyStats.recovering} (${((ecologyStats.recovering / ecologyStats.total) * 100).toFixed(1)}%)`);
  console.log(`  Good (40+): ${ecologyStats.good} (${((ecologyStats.good / ecologyStats.total) * 100).toFixed(1)}%) ← FIX WORKING!`);

  if (ecologyStats.scores.length > 0) {
    console.log(`\n  Statistics (non-crashed):`);
    console.log(`    Mean: ${ecologyStats.mean.toFixed(2)}`);
    console.log(`    Median: ${ecologyStats.median.toFixed(2)}`);
    console.log(`    Std Dev: ${ecologyStats.stdDev.toFixed(2)}`);
    console.log(`    Range: ${Math.min(...ecologyStats.scores).toFixed(2)} - ${Math.max(...ecologyStats.scores).toFixed(2)}`);
  }

  console.log('\n## Paradigm Score Analysis');
  console.log(`  Western Liberal: ${paradigmStats.westernLiberal.nullCount} null / ${paradigmStats.westernLiberal.values.length} valid`);
  if (paradigmStats.westernLiberal.values.length > 0) {
    const mean = paradigmStats.westernLiberal.values.reduce((a, b) => a + b, 0) / paradigmStats.westernLiberal.values.length;
    console.log(`    Mean (valid): ${mean.toFixed(2)}`);
  }

  console.log(`  Development: ${paradigmStats.development.nullCount} null / ${paradigmStats.development.values.length} valid`);
  if (paradigmStats.development.values.length > 0) {
    const mean = paradigmStats.development.values.reduce((a, b) => a + b, 0) / paradigmStats.development.values.length;
    console.log(`    Mean (valid): ${mean.toFixed(2)}`);
  }

  console.log(`  Ecological: ${paradigmStats.ecological.nullCount} null / ${paradigmStats.ecological.values.length} valid`);
  // Already covered above

  console.log(`  Indigenous: ${paradigmStats.indigenous.nullCount} null / ${paradigmStats.indigenous.values.length} valid`);
  if (paradigmStats.indigenous.values.length > 0) {
    const mean = paradigmStats.indigenous.values.reduce((a, b) => a + b, 0) / paradigmStats.indigenous.values.length;
    console.log(`    Mean (valid): ${mean.toFixed(2)}`);
  }

  if (crashedRuns.length > 0) {
    console.log(`\n## Crashed Runs (${crashedRuns.length})`);
    crashedRuns.slice(0, 10).forEach(run => {
      console.log(`  Seed ${run.seed}: crashed at month ${run.finalMonth}`);
    });
    if (crashedRuns.length > 10) {
      console.log(`  ... and ${crashedRuns.length - 10} more`);
    }
  }

  console.log('\n## Key Findings');
  if (ecologyStats.good > 0) {
    console.log(`  ✅ Ecology fix IS WORKING in ${ecologyStats.good}/${ecologyStats.total} runs (${((ecologyStats.good / ecologyStats.total) * 100).toFixed(1)}%)`);
  }
  if (ecologyStats.collapse > 0) {
    console.log(`  ⚠️  Ecology still collapsing in ${ecologyStats.collapse}/${ecologyStats.total} runs (${((ecologyStats.collapse / ecologyStats.total) * 100).toFixed(1)}%)`);
  }
  if (ecologyStats.crashed > 0) {
    console.log(`  ❌ ${ecologyStats.crashed} runs crashed with null ecology scores`);
  }
  if (paradigmStats.westernLiberal.nullCount === files.length) {
    console.log(`  ❌ Western Liberal ALL NULL - calculation bug!`);
  }
  if (outcomes['dystopia'] === files.length) {
    console.log(`  ⚠️  100% dystopia despite ecology improvements - outcome calculation issue?`);
  }

  console.log('');
}

// Run analysis
const outputDir = path.join(process.cwd(), 'monteCarloOutputs');
analyzeRuns(outputDir);

File: scripts/analyze_mc_results.py
#!/usr/bin/env python3
"""
Monte Carlo Results Analysis for RD-1 and RD-3 Validation
Analyzes bifurcation metrics and simulation outcomes across N=10 runs
"""

import json
import os
import numpy as np
from pathlib import Path
from collections import Counter

def analyze_bifurcation_metrics():
    """Analyze bifurcation metrics files for seeds 42000-42009"""
    output_dir = Path(__file__).parent.parent / "monteCarloOutputs"

    results = []
    for seed in range(42000, 42010):
        filepath = output_dir / f"bifurcation_metrics_seed{seed}.json"
        if filepath.exists():
            with open(filepath) as f:
                data = json.load(f)
                results.append(data)

    if not results:
        print("❌ No bifurcation metrics files found for seeds 42000-42009")
        return None

    print(f"\n{'='*80}")
    print(f"MONTE CARLO VALIDATION: RD-1 Permafrost + RD-3 Geopolitical Conflict")
    print(f"{'='*80}\n")
    print(f"Runs analyzed: N={len(results)}")
    print(f"Seeds: {results[0]['seed']} - {results[-1]['seed']}\n")

    # 1. DETERMINISM CHECK
    print(f"\n{'='*80}")
    print("1. DETERMINISM CHECK")
    print(f"{'='*80}\n")

    outcomes = [r['outcome'] for r in results]
    final_pops = [r['finalPopulation'] for r in results]
    final_qols = [r['finalQOL'] for r in results]

    # Check for identical outcomes (perfect determinism with same seed not expected across different seeds)
    outcome_counts = Counter(outcomes)
    print(f"Outcome distribution:")
    for outcome, count in sorted(outcome_counts.items()):
        pct = 100 * count / len(results)
        print(f"  {outcome}: {count}/{len(results)} ({pct:.1f}%)")

    # Check variance in numerical outputs (should be high across different seeds)
    pop_cv = 100 * np.std(final_pops) / np.mean(final_pops) if np.mean(final_pops) > 0 else 0
    qol_cv = 100 * np.std(final_qols) / np.mean(final_qols) if np.mean(final_qols) > 0 else 0

    print(f"\nFinal Population:")
    print(f"  Mean: {np.mean(final_pops):.4f} billion")
    print(f"  Std Dev: {np.std(final_pops):.4f}")
    print(f"  CV: {pop_cv:.2f}%")
    print(f"  Range: [{np.min(final_pops):.4f}, {np.max(final_pops):.4f}]")

    print(f"\nFinal QoL:")
    print(f"  Mean: {np.mean(final_qols):.4f}")
    print(f"  Std Dev: {np.std(final_qols):.4f}")
    print(f"  CV: {qol_cv:.2f}%")
    print(f"  Range: [{np.min(final_qols):.4f}, {np.max(final_qols):.4f}]")

    # Note: Different seeds SHOULD produce different outcomes (that's the point of Monte Carlo)
    print(f"\n✅ Different RNG seeds producing different outcomes (expected behavior)")

    # 2. BIFURCATION ANALYSIS
    print(f"\n{'='*80}")
    print("2. BIFURCATION & TIPPING POINT ANALYSIS")
    print(f"{'='*80}\n")

    bifurcation_types = {
        'environmental': [],
        'social': [],
        'economic': [],
        'governance': [],
        'technology': [],
        'flourishing': []
    }

    for result in results:
        for bif_type, bif_data in result['bifurcations'].items():
            bifurcation_types[bif_type].append(bif_data)

    for bif_type, occurrences in bifurcation_types.items():
        occurred_count = sum(1 for b in occurrences if b.get('occurred', False))
        occurrence_rate = 100 * occurred_count / len(occurrences)

        months = [b.get('month', 0) for b in occurrences if b.get('occurred', False)]
        avg_month = np.mean(months) if months else 0

        print(f"{bif_type.upper()}:")
        print(f"  Occurrence rate: {occurred_count}/{len(occurrences)} ({occurrence_rate:.1f}%)")
        if months:
            print(f"  Average timing: Month {avg_month:.1f}")
            print(f"  Range: Months {min(months)}-{max(months)}")
        print()

    # Regime shift analysis
    print(f"\n{'='*80}")
    print("3. REGIME SHIFT EVENTS")
    print(f"{'='*80}\n")

    all_regime_shifts = []
    for result in results:
        all_regime_shifts.extend(result.get('regimeShiftEvents', []))

    if all_regime_shifts:
        shift_by_system = {}
        for shift in all_regime_shifts:
            system = shift['system']
            if system not in shift_by_system:
                shift_by_system[system] = []
            shift_by_system[system].append(shift)

        for system, shifts in sorted(shift_by_system.items()):
            months = [s['month'] for s in shifts]
            amplifications = [s['amplification'] for s in shifts]

            print(f"{system.upper()} regime shifts:")
            print(f"  Count: {len(shifts)} events across {len(results)} runs")
            print(f"  Average month: {np.mean(months):.1f}")
            print(f"  Average amplification: {np.mean(amplifications):.2f}×")
            print(f"  Max amplification: {np.max(amplifications):.2f}×")
            print()

    # 4. VARIANCE AMPLIFICATION
    print(f"\n{'='*80}")
    print("4. CRITICAL SLOWING DOWN (Variance Amplification)")
    print(f"{'='*80}\n")

    max_amplifications = [r.get('maxVarianceAmplification', 0) for r in results]
    print(f"Maximum variance amplification:")
    print(f"  Mean: {np.mean(max_amplifications):.2f}×")
    print(f"  Std Dev: {np.std(max_amplifications):.2f}×")
    print(f"  Range: [{np.min(max_amplifications):.2f}×, {np.max(max_amplifications):.2f}×]")

    if np.mean(max_amplifications) > 5:
        print(f"\n⚠️  HIGH variance amplification detected (>{5:.1f}×)")
        print(f"     Indicates critical slowing down near tipping points")

    # 5. DISTANCE TO THRESHOLDS
    avg_distances = [r.get('avgDistanceToThresholds', 0) for r in results]
    print(f"\nAverage distance to critical thresholds:")
    print(f"  Mean: {np.mean(avg_distances):.6f}")
    print(f"  Std Dev: {np.std(avg_distances):.6f}")

    if np.mean(avg_distances) < 0.01:
        print(f"\n⚠️  VERY CLOSE to critical thresholds (< 0.01)")
        print(f"     System is near tipping points")

    # SUMMARY STATISTICS
    print(f"\n{'='*80}")
    print("5. SUMMARY & ASSESSMENT")
    print(f"{'='*80}\n")

    # Crash rate
    crash_rate = 100 * sum(1 for r in results if r.get('crashed', False)) / len(results)
    print(f"Crash rate: {crash_rate:.1f}%")

    # Most common outcome
    most_common_outcome = outcome_counts.most_common(1)[0]
    print(f"Most common outcome: {most_common_outcome[0]} ({100*most_common_outcome[1]/len(results):.1f}%)")

    # Overall stability assessment
    print(f"\nSTABILITY ASSESSMENT:")

    issues = []
    if crash_rate > 0:
        issues.append(f"CRITICAL: {crash_rate:.1f}% crash rate")

    if np.mean(max_amplifications) > 10:
        issues.append(f"HIGH: Extreme variance amplification ({np.mean(max_amplifications):.1f}×)")

    if outcome_counts.get('DYSTOPIA', 0) > len(results) * 0.8:
        issues.append(f"WARNING: {100*outcome_counts.get('DYSTOPIA', 0)/len(results):.1f}% dystopia rate")

    if issues:
        print("  ❌ ISSUES DETECTED:")
        for issue in issues:
            print(f"     - {issue}")
    else:
        print("  ✅ No critical issues detected")

    return results

if __name__ == "__main__":
    results = analyze_bifurcation_metrics()

    if results:
        print(f"\n{'='*80}")
        print(f"Analysis complete. Bifurcation metrics only available.")
        print(f"For RD-1 (Permafrost) and RD-3 (Geopolitical) specific validation,")
        print(f"need to examine detailed simulation logs or add custom metrics.")
        print(f"{'='*80}\n")
